tls/datadog_checks/tls/tls.py

```diff
@@ -0,0 +1,353 @@
+            self._cert = expanduser(self._cert)
+
+        self._private_key = self.instance.get('private_key')
+        if self._private_key:
+            self._private_key = expanduser(self._private_key)
+
+        self._cafile = None
+        self._capath = None
+        ca_cert = self.instance.get('ca_cert')
+        if ca_cert:
+            ca_cert = expanduser(ca_cert)
+            if isdir(ca_cert):
+                self._capath = ca_cert
+            else:
+                self._cafile = ca_cert
+
+        # Thresholds expressed in seconds take precedence over those expressed in days
+        self._seconds_warning = (
+            int(self.instance.get('seconds_warning', 0))
+            or days_to_seconds(float(self.instance.get('days_warning', 0)))
+            or self.DEFAULT_EXPIRE_SECONDS_WARNING
+        )
+        self._seconds_critical = (
+            int(self.instance.get('seconds_critical', 0))
+            or days_to_seconds(float(self.instance.get('days_critical', 0)))
+            or self.DEFAULT_EXPIRE_SECONDS_CRITICAL
+        )
+
+        # https://docs.python.org/3/library/ssl.html#ssl.SSLSocket.version
+        self._allowed_versions = get_protocol_versions(self.instance.get('allowed_versions', []))
+
+        # Global tags
+        self._tags = self.instance.get('tags', [])
+        if self._name:
+            self._tags.append('name:{}'.format(self._name))
+
+        # Decide the method of collection for this instance (local file vs remote connection)
+        if self._local_cert_path:
+            self.check = self.check_local
+            if self._validate_hostname and self._server_hostname:
+                self._tags.append('server_hostname:{}'.format(self._server_hostname))
+        else:
+            self.check = self.check_remote
+            self._tags.append('server_hostname:{}'.format(self._server_hostname))
+            self._tags.append('server:{}'.format(self._server))
+            self._tags.append('port:{}'.format(self._port))
+
+        # Assign lazily since these aren't used by both collection methods
+        self._validation_data = None
+        self._local_cert_loader = None
+        self._tls_context = None
+
+    def check_remote(self, instance):
+        if not self._server:
+            raise ConfigurationError('You must specify `server` in your configuration file.')
+
+        try:
+            sock = self.create_connection()
+        except Exception as e:
+            self.service_check(self.SERVICE_CHECK_CAN_CONNECT, self.CRITICAL, tags=self._tags, message=str(e))
+            return
+        else:
+            self.service_check(self.SERVICE_CHECK_CAN_CONNECT, self.OK, tags=self._tags)
+
+        # Get the cert & TLS version from the connection
+        with closing(sock):
+            try:
+                with closing(self.tls_context.wrap_socket(sock, server_hostname=self._server_hostname)) as secure_sock:
+                    der_cert = secure_sock.getpeercert(binary_form=True)
```

Comment:
Does this not automatically perform validations about the hostname for you?

Reply:
I was under the impression that you'd receive an error at this step if the server_hostname didn't match server.

Reply:
https://github.com/DataDog/integrations-core/blob/d66dd62322141ac1df8a6446803aecbb6f08fda7/tls/datadog_checks/tls/tls.py#L333-L335

Reply:
Oh interesting. Does this only do server name validation or everything? Wondering if there are other scenarios, like what we've run into with the http_check that some SSL Error would be thrown here?

Reply:
Just server name validation https://en.wikipedia.org/wiki/Server_Name_Indication